src/main/java/ro/msg/learning/shop/service/OrderService.java

```diff
@@ -21,62 +28,70 @@ public class OrderService {
     private final OrderRepository orderRepository;
     private final ProductRepository productRepository;
     private final StockRepository stockRepository;
+    private final OrderDetailsRepository orderDetailsRepository;
 
 
-    public Order createOrder(OrderDto orderDto) {
+    public Order createOrder(Order order, List<ProductQuantityDto> products) {
 
-        List<StockDto> availableStocks = findLocationWithStock(orderDto);
+        List<StockDto> availableStocks = findLocationWithStock(products);
 
         if (availableStocks.isEmpty()) {
             throw new NoStocksAvailableException();
         }
 
-        orderDto.getProducts().forEach(p -> {
-            availableStocks.forEach(s -> updateStock(s, p.getQuantity()));
+        Set<OrderDetail> orderDetails = products.stream()
+                .map(p -> new OrderDetail(productRepository.findById(p.getProductId()).get(), p.getQuantity()))
+                .collect(Collectors.toSet());
+
+        products.forEach(p -> {
+            availableStocks.forEach(s -> {
+                if (p.getProductId().equals(s.getProduct().getId())) {
+                    updateStock(s, p.getQuantity());
```

Comment:
You should update the stocks only after Save Order was performed successfully. If you update the stock first and then the save order operation fails, you will have to revert the update stock operation. By the way, you can use @Transactional annotation to assure the atomicity of a transaction.